politeiawww/totp.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+package main
+
+import (
+	"bytes"
+	"image/png"
+	"time"
+
+	www "github.com/decred/politeia/politeiawww/api/www/v1"
+	"github.com/decred/politeia/politeiawww/user"
+	"github.com/pquerna/otp/totp"
+)
+
+// processSetTOTP attempts to set a new TOTP key based on the given TOTP type.
+func (p *politeiawww) processSetTOTP(st www.SetTOTP, u *user.User) (*www.SetTOTPReply, error) {
+	log.Tracef("processSetTOTP: %v", u.ID.String())
+	// if the user already has a TOTP secret set, check the code that was given
+	// as well to see if it matches to update.
+	if u.TOTPSecret != "" && u.TOTPVerified {
+		valid := totp.Validate(st.Code, u.TOTPSecret)
+		if !valid {
+			return nil, www.UserError{
+				ErrorCode: www.ErrorStatusTOTPFailedValidation,
+			}
+		}
+	}
+
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      p.cfg.Mode,
```

Comment:
I'm wary of using the mode here.  As of right now, the mode is just a config variable that dictates what routes get registered.  As politeia evolves and as the plugin system evolves, the modes might change, but this locks us into the existing modes.  I think it would be better here to use a const like "decred" here so that it doesn't impact other parts of the code.

Reply:
Is this field ever shown to the user?  If so then we should probably use application specific consts, "politeia" "cms".